senders/msteams/msteams.go

```diff
@@ -0,0 +1,219 @@
+package msteams
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/moira-alert/moira"
+	"github.com/russross/blackfriday/v2"
+)
+
+// Sender implements moira sender interface via MS Teams
+type Sender struct {
+	frontURI string
+	logger   moira.Logger
+	location *time.Location
+	client   *http.Client
+}
+
+// Init initialises settings required for full functionality
+func (sender *Sender) Init(senderSettings map[string]string, logger moira.Logger, location *time.Location, dateTimeFormat string) error {
+	sender.logger = logger
+	sender.location = location
+	sender.frontURI = senderSettings["front_uri"]
+	sender.client = &http.Client{
+		Timeout: time.Duration(30) * time.Second,
+	}
+	return nil
+}
+
+// SendEvents implements Sender interface Send
+func (sender *Sender) SendEvents(events moira.NotificationEvents, contact moira.ContactData, trigger moira.TriggerData, plot []byte, throttled bool) error {
+
+	err := sender.isValidWebhookURL(contact.Value)
+	if err != nil {
+		return fmt.Errorf("invalid msteams webhook url: %s", err.Error())
+	}
+
+	request, err := sender.buildRequest(events, contact, trigger, plot, throttled)
+
+	if err != nil {
+		return fmt.Errorf("failed to build request: %s", err.Error())
+	}
+
+	response, err := sender.client.Do(request)
+
+	if err != nil {
+		return fmt.Errorf("failed to perform request: %s", err.Error())
+	}
+	defer response.Body.Close()
+
+	// read the entire response as required by https://golang.org/pkg/net/http/#Client.Do
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Errorf("failed to decode response: %s", err.Error())
+	}
+
+	//handle non 2xx responses
+	if response.StatusCode >= http.StatusBadRequest && response.StatusCode <= http.StatusNetworkAuthenticationRequired {
+		return fmt.Errorf("server responded with a non 2xx code: %d", response.StatusCode)
+	}
+
+	responseData := string(body)
+	if responseData != "1" {
+		return fmt.Errorf("teams endpoint responded with an error: %s", errors.New(responseData))
+	}
+
+	return nil
+}
+
+func (sender *Sender) buildMessage(events moira.NotificationEvents, trigger moira.TriggerData, throttled bool) MessageCard {
+
+	title, uri := sender.buildTitleAndURI(events, trigger)
+	var triggerDescription string
+	if trigger.Desc != "" {
+		triggerDescription = string(blackfriday.Run([]byte(trigger.Desc)))
+	}
+	facts := sender.buildEventsFacts(events, -1, throttled)
```

Comment:
What is the purpose to set `maxEvents` here as hardcoded magic value? I believe that this value should be set in configuration

Reply:
aaah I fell through the D'OH tree and hit every branch on the way down, thanks for spotting this.